finality-aleph/src/new_network/connection_manager/discovery.rs

```diff
@@ -0,0 +1,690 @@
+    next_query: NodeIndex,
+}
+
+fn authentication_broadcast(authentication: Authentication) -> (DiscoveryMessage, DataCommand) {
+    (
+        DiscoveryMessage::AuthenticationBroadcast(authentication),
+        DataCommand::Broadcast,
+    )
+}
+
+fn request(
+    missing_authorities: Vec<NodeIndex>,
+    authentication: Authentication,
+    peer_id: PeerId,
+) -> (DiscoveryMessage, DataCommand) {
+    (
+        DiscoveryMessage::Request(missing_authorities, authentication),
+        DataCommand::SendTo(peer_id, Protocol::Generic),
+    )
+}
+
+fn response(
+    authentications: Vec<Authentication>,
+    peer_id: PeerId,
+) -> (DiscoveryMessage, DataCommand) {
+    (
+        DiscoveryMessage::Authentications(authentications),
+        DataCommand::SendTo(peer_id, Protocol::Generic),
+    )
+}
+
+impl Discovery {
+    /// Create a new discovery handler with the given response/broadcast cooldown.
+    pub fn new(cooldown: time::Duration) -> Self {
+        Discovery {
+            cooldown,
+            last_broadcast: HashMap::new(),
+            last_response: HashMap::new(),
+            requested_authentications: HashMap::new(),
+            next_query: NodeIndex(0),
+        }
+    }
+
+    /// Returns messages that should be sent as part of authority discovery at this moment.
+    pub fn discover_authorities(
+        &mut self,
+        handler: &SessionHandler,
+    ) -> Vec<(DiscoveryMessage, DataCommand)> {
+        let missing_authorities = handler.missing_nodes();
+        if missing_authorities.is_empty() {
+            return Vec::new();
+        }
+        let node_count = handler.node_count();
+        let authentication = handler.authentication();
+        if missing_authorities.len() * 3 > 2 * node_count.0 {
+            // We know of fewer than 1/3 authorities, broadcast our authentication and hope others
+            // respond in kind.
+            vec![authentication_broadcast(authentication)]
+        } else {
+            // Attempt learning about more authorities from the ones you already know.
+            let mut result = Vec::new();
+            let mut target = self.next_query;
+            while result.len() < 2 {
+                if let Some(peer_id) = handler.peer_id(&target) {
+                    result.push(request(
+                        missing_authorities.clone(),
+                        authentication.clone(),
+                        peer_id,
+                    ));
+                }
+                target = NodeIndex((target.0 + 1) % node_count.0);
+            }
+            self.next_query = target;
+            result
+        }
+    }
+
+    fn handle_authentication(
+        &mut self,
+        authentication: Authentication,
+        handler: &mut SessionHandler,
+    ) -> Vec<Multiaddr> {
+        if !handler.handle_authentication(authentication.clone()) {
+            return Vec::new();
+        }
+        authentication.0.address()
+    }
+
+    fn handle_broadcast(
```

Comment:
Whoops, you are right, we should, I thought it wasn't necessary, but there is a very strange failure mode possible here. I will think about this a bit before fixing it, as it might require changing all the return types, which would be quite annoying...

Reply:
Could you please explain this failure mode?

Reply:
Phew, done by changing the responses into `Vec`s. Only slightly annoying and probably actually better this way anyway. <_<"